pants fails when trying to install plugins that depend on python C externsions

We are seeing this in CI:
```
building 'pyhdrh' extension
creating build/temp.linux-x86_64-3.9
creating build/temp.linux-x86_64-3.9/src
clang -pthread -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -I/tools/deps/include -I/tools/deps/include/ncursesw -fPIC -I/home/toolchain/.cache/pants/named_caches/pex_root/venvs/29638d5feb42d02ca1bfbcc777675dd78fc662d7/ea591207bad3334b92d74c8d0207c58824a6d437/include -I/home/toolchain/.cache/nce/cdc3a4cfddcd63b6cebdd75b14970e02d8ef0ac5be4d350e57ab5df56c19e85e/cpython-3.9.15+20221106-x86_64-unknown-linux-gnu-install_only.tar.gz/python/include/python3.9 -c src/python-codec.c -o build/temp.linux-x86_64-3.9/src/python-codec.o
error: command 'clang' failed: No such file or directory
----------------------------------------
ERROR: Failed building wheel for hdrhistogram
```

we use this package [hdrhistogram](https://pypi.org/project/hdrhistogram/) in our python code.

Ok, the full error was:
```
20:07:23.94 [INFO] [rule-construct-auth-store] Successfully loaded Toolchain token from env var 'TOOLCHAIN_AUTH_TOKEN', expiration: 2024-01-05T18:38:56+00:00.
20:07:28.99 [INFO] View on BuildSense: https://app.toolchain.com/organizations/toolchainlabs/repos/toolchain/builds/pants_run_2023_01_24_20_07_23_798_6a9d85271d104289ba1839915c08b115/
20:07:33.57 [WARN] The lint goal was called with targets including `src/python/toolchain/aws/acm_test.py:tests`, `src/python/toolchain/aws/aws_api_test.py:tests`, `src/python/toolchain/aws/dynamodb_test.py:tests` (and others), which specify the environment `remote`, which is a `remote_environment`. The lint goal only runs in the local environment. You may experience unexpected behavior.
20:08:29.12 [INFO] Completed: Lint with regex patterns - regex-lint succeeded.
38 files matched all required patterns.
0 files failed to match at least one required pattern.
20:08:29.12 [INFO] Completed: Lint with regex patterns - regex-lint succeeded.
13 files matched all required patterns.
0 files failed to match at least one required pattern.
20:08:29.12 [INFO] Completed: Lint with regex patterns - regex-lint succeeded.
26 files matched all required patterns.
0 files failed to match at least one required pattern.
20:08:29.12 [INFO] Completed: Lint with regex patterns - regex-lint succeeded.
2 files matched all required patterns.
0 files failed to match at least one required pattern.
20:08:29.13 [INFO] Completed: Lint with regex patterns - regex-lint succeeded.
71 files matched all required patterns.
0 files failed to match at least one required pattern.
20:08:29.14 [INFO] Completed: Lint with regex patterns - regex-lint succeeded.
13 files matched all required patterns.
0 files failed to match at least one required pattern.
20:08:29.14 [INFO] Completed: Lint with regex patterns - regex-lint succeeded.
38 files matched all required patterns.
0 files failed to match at least one required pattern.
20:08:29.14 [INFO] Completed: Lint with regex patterns - regex-lint succeeded.
46 files matched all required patterns.
0 files failed to match at least one required pattern.
20:08:29.14 [INFO] Completed: Lint with regex patterns - regex-lint succeeded.
34 files matched all required patterns.
0 files failed to match at least one required pattern.
20:08:29.14 [INFO] Completed: Lint with regex patterns - regex-lint succeeded.
10 files matched all required patterns.
0 files failed to match at least one required pattern.
20:08:29.18 [INFO] Completed: Lint with regex patterns - regex-lint succeeded.
65 files matched all required patterns.
0 files failed to match at least one required pattern.
20:08:29.18 [INFO] Completed: Lint with regex patterns - regex-lint succeeded.
74 files matched all required patterns.
0 files failed to match at least one required pattern.
20:08:29.18 [INFO] Completed: Lint with regex patterns - regex-lint succeeded.
53 files matched all required patterns.
0 files failed to match at least one required pattern.
20:08:29.18 [INFO] Completed: Lint with regex patterns - regex-lint succeeded.
48 files matched all required patterns.
0 files failed to match at least one required pattern.
20:08:29.18 [INFO] Completed: Lint with regex patterns - regex-lint succeeded.
82 files matched all required patterns.
0 files failed to match at least one required pattern.
20:08:29.18 [INFO] Completed: Lint with regex patterns - regex-lint succeeded.
32 files matched all required patterns.
0 files failed to match at least one required pattern.
20:08:29.19 [INFO] Completed: Lint with regex patterns - regex-lint succeeded.
87 files matched all required patterns.
0 files failed to match at least one required pattern.
20:08:29.19 [INFO] Completed: Lint with regex patterns - regex-lint succeeded.
93 files matched all required patterns.
0 files failed to match at least one required pattern.
20:08:29.19 [INFO] Completed: Lint with regex patterns - regex-lint succeeded.
130 files matched all required patterns.
0 files failed to match at least one required pattern.
20:08:29.19 [INFO] Completed: Lint with regex patterns - regex-lint succeeded.
139 files matched all required patterns.
0 files failed to match at least one required pattern.
20:08:29.20 [INFO] Completed: Lint with regex patterns - regex-lint succeeded.
216 files matched all required patterns.
0 files failed to match at least one required pattern.
20:08:29.20 [INFO] Completed: Lint with regex patterns - regex-lint succeeded.
160 files matched all required patterns.
0 files failed to match at least one required pattern.
20:08:29.21 [INFO] Completed: Lint with regex patterns - regex-lint succeeded.
122 files matched all required patterns.
0 files failed to match at least one required pattern.
20:08:29.25 [ERROR] Completed: Lint with regex patterns - regex-lint failed (exit code 1).
X src/sh/pants/pantsup.sh: Didn't match: shell_header
64 files matched all required patterns.
1 files failed to match at least one required pattern.
20:08:29.32 [INFO] Completed: Lint with regex patterns - regex-lint succeeded.
381 files matched all required patterns.
0 files failed to match at least one required pattern.
20:08:29.33 [INFO] Completed: pants.backend.go.lint.gofmt.rules.gofmt_fmt - gofmt made no changes.
20:08:29.59 [INFO] Starting: Building bandit.pex from 3rdparty/python/bandit.lock
20:08:29.59 [INFO] Starting: Building bandit.pex from 3rdparty/python/bandit.lock
20:08:31.19 [INFO] Starting: Building flake8.pex from 3rdparty/python/flake8.lock
20:08:31.74 [INFO] Starting: Building pyupgrade.pex from 3rdparty/python/pyupgrade.lock
20:08:32.38 [ERROR] The scheduler was invalidated: Exception('pantsd process 3631 was using 1033.99 MiB of memory (above the `--pantsd-max-memory-usage` limit of 1024.00 MiB).')
20:08:32.51 [ERROR] service failure for <pants.pantsd.service.scheduler_service.SchedulerService object at 0x7f53fbe25dc0>.
20:08:32.58 [INFO] Completed: Format with shfmt - shfmt made no changes.
20:08:32.58 [INFO] Waiting for ongoing runs to complete before exiting...
20:08:37.19 [INFO] Completed: Building pyupgrade.pex from 3rdparty/python/pyupgrade.lock
20:08:37.19 [INFO] Starting: Building docformatter.pex from docformatter_default.lock
20:08:41.18 [INFO] Completed: Building bandit.pex from 3rdparty/python/bandit.lock
20:08:41.18 [INFO] Starting: Building autoflake.pex from 3rdparty/python/autoflake.lock
20:08:41.40 [INFO] Completed: Building bandit.pex from 3rdparty/python/bandit.lock
20:08:41.40 [INFO] Starting: Building flake8.pex from 3rdparty/python/flake8.lock
20:08:41.93 [INFO] Completed: Building docformatter.pex from docformatter_default.lock
20:08:41.93 [INFO] Starting: Building isort.pex from 3rdparty/python/isort.lock
20:08:43.98 [INFO] Extending leases
20:08:44.10 [INFO] Completed: Building isort.pex from 3rdparty/python/isort.lock
20:08:44.10 [INFO] Starting: Building pylint.pex from 3rdparty/python/pylint.lock
20:08:44.83 [INFO] Completed: Building autoflake.pex from 3rdparty/python/autoflake.lock
20:08:44.83 [INFO] Starting: Building pylint.pex from 3rdparty/python/pylint.lock
20:08:48.92 [INFO] Done extending leases
20:08:50.70 [INFO] Completed: Building flake8.pex from 3rdparty/python/flake8.lock
20:08:50.70 [INFO] Starting: Building black.pex from 3rdparty/python/black.lock
20:08:51.20 [INFO] Completed: Building flake8.pex from 3rdparty/python/flake8.lock
20:08:51.20 [INFO] Starting: Building 3 requirements for requirements.pex from the 3rdparty/python/pants_plugin.lock resolve: chardet==5.0.0, pantsbuild.pants<2.16,>=2.15.0a0, requests>=2.28.1
20:08:56.93 [INFO] Completed: Building black.pex from 3rdparty/python/black.lock
20:08:56.93 [INFO] Starting: Building 42 requirements for requirements.pex from the 3rdparty/python/default_toolchain.lock resolve: Django==4.1.5; python_version >= "3.8", ansicolors==1.1.8, beautifulsoup4==4.11.1, boto3==1.26.54... (793 characters truncated)
20:09:01.02 [INFO] Completed: Building pylint.pex from 3rdparty/python/pylint.lock
20:09:01.02 [INFO] Starting: Building 49 requirements for requirements.pex from the 3rdparty/python/default_toolchain.lock resolve: Django==4.1.5; python_version >= "3.8", ansicolors==1.1.8, boto3==1.26.54, botocore, certifi==202... (909 characters truncated)
20:09:01.50 [INFO] Completed: Building pylint.pex from 3rdparty/python/pylint.lock
20:09:01.50 [INFO] Starting: Building 11 requirements for requirements.pex from the 3rdparty/python/pants_plugin.lock resolve: chardet==5.0.0, defusedxml, faker, freezegun, pantsbuild.pants.testutil<2.16,>=2.15.0a0, pantsbuild.pa... (102 characters truncated)
20:09:25.47 [INFO] Completed: Building 3 requirements for requirements.pex from the 3rdparty/python/pants_plugin.lock resolve: chardet==5.0.0, pantsbuild.pants<2.16,>=2.15.0a0, requests>=2.28.1
20:09:25.48 [INFO] Starting: Building 38 requirements for requirements.pex from the 3rdparty/python/default_toolchain.lock resolve: Django==4.1.5; python_version >= "3.8", ansicolors==1.1.8, boto3==1.26.54, botocore, django-exten... (698 characters truncated)
20:09:40.81 [INFO] Completed: Building 38 requirements for requirements.pex from the 3rdparty/python/default_toolchain.lock resolve: Django==4.1.5; python_version >= "3.8", ansicolors==1.1.8, boto3==1.26.54, botocore, django-exten... (698 characters truncated)
20:09:40.81 [INFO] Starting: Building 44 requirements for requirements.pex from the 3rdparty/python/default_toolchain.lock resolve: Django==4.1.5; python_version >= "3.8", ansicolors==1.1.8, boto3==1.26.54, botocore, cryptography... (865 characters truncated)
20:09:45.26 [INFO] Completed: Building 11 requirements for requirements.pex from the 3rdparty/python/pants_plugin.lock resolve: chardet==5.0.0, defusedxml, faker, freezegun, pantsbuild.pants.testutil<2.16,>=2.15.0a0, pantsbuild.pa... (102 characters truncated)
20:09:45.26 [INFO] Starting: Building 21 requirements for requirements.pex from the 3rdparty/python/default_toolchain.lock resolve: Django==4.1.5; python_version >= "3.8", ansicolors==1.1.8, boto3==1.26.54, botocore, django-exten... (388 characters truncated)
20:09:45.90 [INFO] Completed: Building 42 requirements for requirements.pex from the 3rdparty/python/default_toolchain.lock resolve: Django==4.1.5; python_version >= "3.8", ansicolors==1.1.8, beautifulsoup4==4.11.1, boto3==1.26.54... (793 characters truncated)
20:09:45.90 [INFO] Starting: Building 40 requirements for requirements.pex from the 3rdparty/python/default_toolchain.lock resolve: Django==4.1.5; python_version >= "3.8", ansicolors==1.1.8, boto3==1.26.54, botocore, defusedxml==... (757 characters truncated)
20:09:46.16 [INFO] Completed: Building 44 requirements for requirements.pex from the 3rdparty/python/default_toolchain.lock resolve: Django==4.1.5; python_version >= "3.8", ansicolors==1.1.8, boto3==1.26.54, botocore, cryptography... (865 characters truncated)
20:09:46.16 [INFO] Starting: Building 45 requirements for requirements.pex from the 3rdparty/python/default_toolchain.lock resolve: Django==4.1.5; python_version >= "3.8", ansicolors==1.1.8, beautifulsoup4==4.11.1, boto3==1.26.54... (827 characters truncated)
20:09:46.29 [INFO] Canceled: Building 21 requirements for requirements.pex from the 3rdparty/python/default_toolchain.lock resolve: Django==4.1.5; python_version >= "3.8", ansicolors==1.1.8, boto3==1.26.54, botocore, django-exten... (388 characters truncated)
20:09:46.33 [INFO] Canceled: Building 45 requirements for requirements.pex from the 3rdparty/python/default_toolchain.lock resolve: Django==4.1.5; python_version >= "3.8", ansicolors==1.1.8, beautifulsoup4==4.11.1, boto3==1.26.54... (827 characters truncated)
20:09:46.33 [INFO] Canceled: Building 40 requirements for requirements.pex from the 3rdparty/python/default_toolchain.lock resolve: Django==4.1.5; python_version >= "3.8", ansicolors==1.1.8, boto3==1.26.54, botocore, defusedxml==... (757 characters truncated)
20:09:46.33 [INFO] Canceled: Building 49 requirements for requirements.pex from the 3rdparty/python/default_toolchain.lock resolve: Django==4.1.5; python_version >= "3.8", ansicolors==1.1.8, boto3==1.26.54, botocore, certifi==202... (909 characters truncated)
20:09:46.41 [ERROR] 1 Exception encountered:
Engine traceback:
in select
..
in pants.core.goals.lint.lint
`lint` goal
in pants.backend.python.lint.pylint.rules.run_pylint
Lint using Pylint
in pants.backend.python.util_rules.pex.create_pex
..
in pants.backend.python.util_rules.pex.build_pex
..
in pants.engine.process.fallible_to_exec_result_or_raise
..
Traceback (most recent call last):
File "/home/toolchain/.cache/nce/94c7754066d3b57746d362573844961b2eb7685b27f101c71b5e163ce0db0253/bindings/venvs/2.15.0rc2/lib/python3.9/site-packages/pants/engine/process.py", line 289, in fallible_to_exec_result_or_raise
raise ProcessExecutionFailure(
pants.engine.process.ProcessExecutionFailure: Process 'Building 44 requirements for requirements.pex from the 3rdparty/python/default_toolchain.lock resolve: Django==4.1.5; python_version >= "3.8", ansicolors==1.1.8, boto3==1.26.54, botocore, cryptography==39.0.0, defusedxml==0.7.1, django-extensions==3.2.1, django-prometheus==2.2.0, django_csp==3.7, djangorestframework==3.14.0, djproxy==2.3.6, drf-nested-routers==0.93.4, elasticsearch-dsl==7.4.0, fasteners, freezegun, gunicorn==20.1.0, hdrhistogram==0.10.1, httpx==0.23.3, humanize==4.4.0, influxdb-client==1.35.0, jinja2==3.1.2, jwcrypto==1.4.2, kubernetes==25.3.0, more-itertools==9.0.0, moto[dynamodb,ec2,s3,ses]==4.1.0, ndjson==0.3.1, opensearch-py==2.0.1, pkginfo==1.9.6, prometheus-client==0.15.0, psycopg2-binary==2.9.4; sys_platform != "darwin", psycopg2==2.9.4; sys_platform == "darwin", pytest-django==4.5.2, pytest-httpx==0.21.3, pytest-responses==0.5.1, pytest==7.2.1, python-dateutil==2.8.2, python-jose[cryptography]==3.3.0, requests-aws4auth==1.2.0, requests==2.28.2, rich==13.2.0, sentry-sdk==1.14.0, setuptools, shortuuid==1.0.11, urllib3==1.26.14' failed with exit code 1.
stdout:
stderr:
pid 7824 -> /home/toolchain/.cache/pants/named_caches/pex_root/venvs/29638d5feb42d02ca1bfbcc777675dd78fc662d7/ea591207bad3334b92d74c8d0207c58824a6d437/bin/python -sE /home/toolchain/.cache/pants/named_caches/pex_root/venvs/29638d5feb42d02ca1bfbcc777675dd78fc662d7/ea591207bad3334b92d74c8d0207c58824a6d437/pex --disable-pip-version-check --no-python-version-warning --exists-action a --no-input --use-deprecated legacy-resolver --isolated -q --cache-dir /home/toolchain/.cache/pants/named_caches/pex_root/pip_cache wheel --no-deps --wheel-dir /home/toolchain/.cache/pants/named_caches/pex_root/built_wheels/sdists/hdrhistogram-0.10.1.tar.gz/542a10f166ea07229cd85a6c09229936749ed7840018993537340fa9ada78566/cp39-cp39-manylinux_2_28_x86_64.cf264f6fd54c48399f6f20f6a2a21dee /home/toolchain/.cache/pants/named_caches/pex_root/downloads/542a10f166ea07229cd85a6c09229936749ed7840018993537340fa9ada78566/hdrhistogram-0.10.1.tar.gz --index-url https://pypi.org/simple/ --retries 5 --timeout 15 exited with 1 and STDERR:
ERROR: Command errored out with exit status 1:
command: /home/toolchain/.cache/pants/named_caches/pex_root/venvs/29638d5feb42d02ca1bfbcc777675dd78fc662d7/ea591207bad3334b92d74c8d0207c58824a6d437/bin/python -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/home/toolchain/.cache/pants/named_caches/pex_root/pip_cache/.tmp/pip-req-build-980w2u2n/setup.py'"'"'; __file__='"'"'/home/toolchain/.cache/pants/named_caches/pex_root/pip_cache/.tmp/pip-req-build-980w2u2n/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /home/toolchain/.cache/pants/named_caches/pex_root/pip_cache/.tmp/pip-wheel-8j6jo0ol
cwd: /home/toolchain/.cache/pants/named_caches/pex_root/pip_cache/.tmp/pip-req-build-980w2u2n/
Complete output (26 lines):
running bdist_wheel
The [wheel] section is deprecated. Use [bdist_wheel] instead.
running build
running build_py
creating build
creating build/lib.linux-x86_64-3.9
creating build/lib.linux-x86_64-3.9/hdrh
copying hdrh/__init__.py -> build/lib.linux-x86_64-3.9/hdrh
copying hdrh/codec.py -> build/lib.linux-x86_64-3.9/hdrh
copying hdrh/dump.py -> build/lib.linux-x86_64-3.9/hdrh
copying hdrh/histogram.py -> build/lib.linux-x86_64-3.9/hdrh
copying hdrh/iterators.py -> build/lib.linux-x86_64-3.9/hdrh
copying hdrh/log.py -> build/lib.linux-x86_64-3.9/hdrh
running egg_info
writing hdrhistogram.egg-info/PKG-INFO
writing dependency_links to hdrhistogram.egg-info/dependency_links.txt
writing entry points to hdrhistogram.egg-info/entry_points.txt
writing requirements to hdrhistogram.egg-info/requires.txt
writing top-level names to hdrhistogram.egg-info/top_level.txt
[pbr] Reusing existing SOURCES.txt
running build_ext
building 'pyhdrh' extension
creating build/temp.linux-x86_64-3.9
creating build/temp.linux-x86_64-3.9/src
clang -pthread -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -I/tools/deps/include -I/tools/deps/include/ncursesw -fPIC -I/home/toolchain/.cache/pants/named_caches/pex_root/venvs/29638d5feb42d02ca1bfbcc777675dd78fc662d7/ea591207bad3334b92d74c8d0207c58824a6d437/include -I/home/toolchain/.cache/nce/cdc3a4cfddcd63b6cebdd75b14970e02d8ef0ac5be4d350e57ab5df56c19e85e/cpython-3.9.15+20221106-x86_64-unknown-linux-gnu-install_only.tar.gz/python/include/python3.9 -c src/python-codec.c -o build/temp.linux-x86_64-3.9/src/python-codec.o
error: command 'clang' failed: No such file or directory
----------------------------------------
ERROR: Failed building wheel for hdrhistogram
ERROR: Failed to build one or more wheels
Use `--keep-sandboxes=on_failure` to preserve the process chroot for inspection.
```

So the issue is in a build step and not in the Pants bootstrap itself. I'm pretty sure this is because scie-panst places its interpreter on the PATH here:
https://github.com/pantsbuild/scie-pants/blob/e8f70ac28f822e6a305c56d9e5d890f1f6e3d159/package/scie-pants.lift.json#L93-L94

That probably should not be done since the scie-pants PBS interpreter is hobbled this way; so that much is an easy fix.

If this would have been a problem in a Pants plugin / during bootstrap though, there would not be an easy solution. I'll break out an issue for that gedanken-bug shortly.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -58,6 +58,13 @@ jobs:
         run: |
           PANTS_BOOTSTRAP_GITHUB_API_BEARER_TOKEN=${{ secrets.GITHUB_TOKEN }} \
             cargo run -p package -- test
+      - name: Setup Python 3.9
+        if: ${{ matrix.os == 'ubuntu-22.04' }}
+        uses: actions/setup-python@v4
+        with:
+          # N.B.: We need Python 3.9 for running Pants goals against our tools.pex Python tools
+          # codebase.
+          python-version: "3.9"
       - name: Build, Package & Integration Tests
         if: ${{ matrix.os == 'ubuntu-22.04' }}
         run: |

--- a/package/scie-pants.lift.json
+++ b/package/scie-pants.lift.json
@@ -91,7 +91,6 @@
           "pants": {
             "description": "Runs a hermetic Pants installation.",
             "env": {
-              "=PATH": "{scie.env.PATH}:{scie.bindings.install:VIRTUAL_ENV}/bin",
               "=PANTS_VERSION": "{scie.bindings.configure:PANTS_VERSION}",
               "PANTS_BUILDROOT_OVERRIDE": "{scie.bindings.configure:PANTS_BUILDROOT_OVERRIDE}"
             },
@@ -104,7 +103,6 @@
           "pants-debug": {
             "description": "Runs a hermetic Pants installation with a debug server for debugging Pants code.",
             "env": {
-              "=PATH": "{scie.env.PATH}:{scie.bindings.install:VIRTUAL_ENV}/bin",
               "=PANTS_VERSION": "{scie.bindings.configure:PANTS_VERSION}",
               "PANTS_BUILDROOT_OVERRIDE": "{scie.bindings.configure:PANTS_BUILDROOT_OVERRIDE}"
             },

